apis/projectcontour/v1/helpers.go

```diff
@@ -78,3 +80,193 @@ func (r *Route) AuthorizationContext(parent map[string]string) map[string]string
 
 	return values
 }
+
+// AddError adds an error-level Subcondition to the DetailedCondition.
+// AddError will also update the DetailedCondition's state to take into account
+// the error that's present.
+// If a SubCondition with the given errorType exists, will change
+// the reason to "MultipleReasons", and will append the given message
+// to the existing SubCondition's message.
+func (dc *DetailedCondition) AddError(errorType, reason, message string) {
+
+	// Update the condition so that it indicates there's at least one error
+	// This needs to be here because conditions may be normal-true (positive)
+	// polarity (like `Valid`), or abnormal-true (negative) polarity
+	// (like `ErrorPresent`)
+	if dc.IsPositivePolarity() {
+		dc.Status = ConditionFalse
+	} else {
+		dc.Status = ConditionTrue
+	}
+
+	// Update the DetailedCondition Reason and message if the reason
+	// is different to the current one. We have to namespace
+	// the DetailedCondition reason by the errorType because we're
+	// adding SubConditions; DetailedCondition Reason is a summary of the
+	// SubCondition reasons.
+	detailedReason := errorType + reason
+	dc.updateReason(detailedReason, message)
+	// Now we can actually upsert the error.
+	i := getIndex(errorType, dc.Errors)
+
+	if i == -1 {
+		dc.Errors = append(dc.Errors, SubCondition{
+			Type:    errorType,
+			Status:  ConditionTrue,
```

Comment:
Yes, all new errors will always be true; in fact all Subconditions will pretty much always be True, because we're reconstructing the Valid condition each time we build a DAG.

Reply:
Also, here we're writing a `SubCondition` status, on line 96, it was the containing `DetailedCondition` status. I'll see if I can make that clearer.